pbjson-types/src/any.rs

```diff
@@ -0,0 +1,248 @@
+const TYPE_FIELD: &str = "@type";
+const VALUE_FIELD: &str = "value";
+
+impl serde::Serialize for crate::Any {
+    #[allow(deprecated)]
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::{SerializeMap, SerializeStruct};
+
+        const NAME: &str = "google.protobuf.Any";
+
+        let value = <crate::Value as prost::Message>::decode(self.value.clone()).map_err(|_| {
```

Comment:
Shouldn't this perform a match on self.type and decode to the corresponding well known type if it matches? This seems to be assuming all `google.protobuf.Any` payloads are encoded as `google.protobuf.Value`?

Reply:
> decode to the corresponding well known type if it matches?

I'm not sure that I follow. If we only decoded matching "well-known" values, then this type can't be used for most projects. For example, in the project I want to use this library in, we're using the [`data-plane-api`](https://github.com/envoyproxy/data-plane-api) which uses `google.protobuf.Any` to store typed configs, those aren't going to be known to this project, so that makes decoding those impossible.

> This seems to be assuming all google.protobuf.Any payloads are encoded as google.protobuf.Value?

Almost correct, it assumes that any payload can be *decoded* as a `google.protobuf.Value`, which to my understanding (which could be wrong), matches any protobuf value.

Reply:
> If we only decoded matching "well-known" values, then this type can't be used for most projects.

Yes, this is a major limitation of `protobuf.Any` and by extension protobuf as a whole. A payload cannot be decoded without knowledge of the schema, in fact the wire format has no concept of the name of a field at all see [here](https://developers.google.com/protocol-buffers/docs/encoding).

> so that makes decoding those impossible.

If the client understands the types it can decode the `value` field to the concrete message type, we do this in IOx for [operations](https://github.com/influxdata/influxdb_iox/blob/3cd7d2eda2d45f2aecae297062cf5c51eb78acd9/generated_types/src/google.rs#L80).

If you wish to handle any protobuf payload, you will need some channel to get the message descriptors (e.g. [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md)) and then a client library that can use these descriptors to decode the messages, but I'm not sure of a Rust implementation of protobuf reflection...

> which to my understanding (which could be wrong), matches any protobuf value.

The wire encoding of a protobuf message is tightly coupled to its schema, whilst any message with the same layout of fields and types could theoretically be decoded as `protobuf.Value`, it would only be correct to decode to a payload that was an actual encoded `protobuf.Value`.